File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-1.java
class Solution {
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> res = new ArrayList<>();

        int rows = heights.length, cols = heights[0].length;

        boolean[][] pacific = new boolean[rows][cols];
        boolean[][] atlantic = new boolean[rows][cols];

        for (int i = 0; i < rows; i++) {
            dfs(-1, i, 0, pacific, heights);
            dfs(-1, i, cols - 1, atlantic, heights);
        }

        for (int j = 0; j < cols; j++) {
            dfs(-1, 0, j, pacific, heights);
            dfs(-1, rows - 1, j, atlantic, heights);
        }

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (pacific[i][j] && atlantic[i][j]) res.add(new ArrayList<>(List.of(i, j)));
            }
        }

        return res;
    }

    private void dfs(int h, int i, int j, boolean[][] visited, int[][] heights) {
        if (i < 0 || j < 0 || i >= heights.length || j >= heights[0].length || visited[i][j] || heights[i][j] < h) return;
        visited[i][j] = true;

        dfs(heights[i][j], i + 1, j, visited, heights);
        dfs(heights[i][j], i - 1, j, visited, heights);
        dfs(heights[i][j], i, j + 1, visited, heights);
        dfs(heights[i][j], i, j - 1, visited, heights);
    }
}

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-1.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        rows, cols = len(matrix), len(matrix[0])
        dp = {}

        def dfs(i, j):
            if (i, j) in dp:
                return dp[(i, j)]
                
            res = 1

            if i + 1 < rows and matrix[i + 1][j] > matrix[i][j]:
                res = max(res, 1 + dfs(i + 1, j))
            if j + 1 < cols and matrix[i][j + 1] > matrix[i][j]:
                res = max(res, 1 + dfs(i, j + 1))
            if i - 1 >= 0 and matrix[i - 1][j] > matrix[i][j]:
                res = max(res, 1 + dfs(i - 1, j))
            if j - 1 >= 0 and matrix[i][j - 1] > matrix[i][j]:
                res = max(res, 1 + dfs(i, j - 1))
            
            dp[(i, j)] = res
            return res

        maxPath = 0
        
        for i in range(rows):
            for j in range(cols):
                maxPath = max(maxPath, dfs(i, j))
        
        return maxPath
